### ISSUE TITLE
Current implementation of Location overwrites states instead of maintaining them.

### ISSUE THREAD

**Issue by kenisteward** [angular/angular#16911]
```
[X] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
When switching from one state to a state you have currently been to using the back button or forward button on the browser the Location class currently overwrites existing history.  This is because when using the platformlocation.replacestate function it sends null instead of checking previously for state that exists.

**Expected behavior**
I expect that if there is any existing state in the brower's History API which you can check by calling history.state that the Location service when doing a replacestate takes that state into consideration before replacing.

**Minimal reproduction of the problem with instructions**
Take any current setup with routing 
Open a page in the browser
add state by doing history.replaceState({}, null);
navigate away from that.
then click back button in browser
check for state by using history.state
notice that the state has been overwritten with null value

**What is the motivation / use case for changing the behavior?**
Angular 2 should be able to support the native History API.  It should not ignore previously set states by script.

**Please tell us about your environment:**
Any browser.

* **Angular version:** 2.0.X
Angular latest

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]
any
 
* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
TypeScript@latest

**Comment by kenisteward** [comment]
as an update I implemented a custom Location class and provided it over the default one with my proposed changes.  I simply changed 

```
  replaceState(path: string, query: string = ''): void {
    this._platformStrategy.replaceState(null, '', path, query);
  }

```
to

```
  replaceState(path: string, query: string = ''): void {
     const currState = window.history.state;
    this._platformStrategy.replaceState(currState, '', path, query);
  }
```
from this code.
https://github.com/angular/angular/blob/master/packages/common/src/location/location.ts

It properly pushed the maintained History API state through on each location pass.   Also if Location is supposed to be the programmatic representation of underlying API's I would suggest adding the ability to set History API through those calls with a change like this.

```
  replaceState(path: string, query: string = '', state?: any): void {
    const stateToUse = state !== undefined ? state : window.history.state
    this._platformStrategy.replaceState(stateToUse, '', path, query);
  }

  go(path: string, query: string = '', state?: any): void {
    this._platformStrategy.pushState(state || null, '', path, query);
  }

```

the idea would be that if state is passed in, we want to use whatever that is over anything that was currently there.  If it wasn't passed in then we can safely use whatever is there.  In the case of go it's fine to just overwrite because we'll be going to a completely new leaf in history.

Note: I realize this may not be the exact way to do it as Location should be completely agnostic of the browser but this is the starting idea i had.

**Comment by Sjonnie2nd** [comment]
I 'd appreciate this adjustment very much.

In AngularJS I am using the state to keep track of back(), forward() and -new path- actions. Only by storing an index in it. In that way I am able to manage some kind of sessionstate, handle fe. dirty formdata, etc..

Please..

**Comment by kenisteward** [comment]
I do this currently by using a custom location class that overwrites the
replace state by checking to see if it exists first

**Comment by jasonaden** [comment]
Looks like this will be fixed with #20030.
